### PR TITLE
Is obstructed

### PR DESCRIPTION
**Pull request by NineDrop**
Sorry its late but I managed to work out the is_obstructed? method. I did have to add a game_id column to the pieces model as well as re migrate some of the columns to the games model, so please makes sure to run a migration when you pull it. I wrote some tests as well for the method and all seems to be working.

### CODE REVIEW

**Comment by NineDrop** [comment]
For the merge conflict, i think we only need to have the pieces belong to the game, rather than have them belong to both the player and the game. The pieces are already associated to a player through the game model, so we don't need to associate them with a player directly.

**Comment by NineDrop** [comment]
For the merge conflict, i think we only need to have the pieces belong to the game, rather than have them belong to both the player and the game. The pieces are already associated to a player through the game model, so we don't need to associate them with a player directly.

File: app/models/piece.rb [hunk 1]
```diff
@@ -1,2 +1,45 @@
 class Piece < ApplicationRecord
+	
+	Pieces.each do |piece|
+		if piece.y_position == start_y && piece.x_position.between?(start_x, dest_x)
+				return false
+			end
+		end
+		return true
+	end
+
```
Comment (liangma87) [inline app/models/piece.rb:12]: @NineDrop it looks that start_y is not defined before it is used. Can you double check?

Comment (liangma87) [reply app/models/piece.rb:12]: @NineDrop does this need to be within a function?

File: app/models/piece.rb [hunk 2]
```diff
@@ -1,2 +1,45 @@
...
+	def long_check?(start_x,start_y,dest_x,dest_y)
+		Pieces.each do |piece|
+			if piece.x_position == start_x && piece.y_position.between?(start_y, dest_y)
+				return false
+			end
+		end
+		return true
+	end
```
Comment (liangma87) [inline app/models/piece.rb:20]: @NineDrop  what is long_check? Suggest to add some comments to brief describe the function.

File: app/models/piece.rb [hunk 3]
```diff
@@ -1,2 +1,45 @@
...
+
+	def diag_check?(start_x,start_y,dest_x,dest_y)
+
+	end
+
+	def valid_move?(start_x,start_y,dest_x,dest_y)
+		if ((start_y - dest_y)/(start_x - dest_x)).abs != (1)
```
Comment (liangma87) [inline app/models/piece.rb]: @NineDrop  We can remove the parenthese around 1. I am not very familiar with chess, why this divide can check whether this is a valid move?

File: app/models/piece.rb [hunk 4]
```diff
@@ -1,2 +1,45 @@
...
+	def valid_move?(start_x,start_y,dest_x,dest_y)
+		if ((start_y - dest_y)/(start_x - dest_x)).abs != (1) 
+			return false
+	end
+
+	def is_obstructed?(start_x,start_y,dest_x,dest_y)
+		start_x = x1
+		start_y = y1
+		dest_x = x2
+		dest_y = y2
```
Comment (liangma87) [inline app/models/piece.rb]: @NineDrop x1, y1, x2, y2 are not defined before using.

File: app/models/piece.rb [hunk 5]
```diff
@@ -1,2 +1,45 @@
...
+		if y1 == y2
+			return lat_check?(x1,y1,x2,y2)
```
Comment (liangma87) [inline app/models/piece.rb]: @NineDrop  is lat_check? defined somewhere else?

**Review by liangma87 — CHANGES_REQUESTED** [review]

**Comment by liangma87** [comment]
@NineDrop will review your other commits later.

**Comment by NineDrop** [comment]
Please review the most recent commit, as that contains the final form of the code.

**Comment by liangma87** [comment]
All,

In case you need it,  a very good article with detailed explanation of the valid chess moves.
https://www.instructables.com/id/Playing-Chess/

Liang

File: app/models/piece.rb (hunk 3 above)
Comment (NineDrop) [reply app/models/piece.rb]: outdated commit

File: app/models/piece.rb (hunk 2 above)
Comment (NineDrop) [reply app/models/piece.rb:20]: outdated commit

File: app/models/piece.rb (hunk 4 above)
Comment (NineDrop) [reply app/models/piece.rb]: outdated commit

File: app/models/piece.rb (hunk 5 above)
Comment (NineDrop) [reply app/models/piece.rb]: outdated commit

File: app/models/piece.rb (hunk 1 above)
Comment (NineDrop) [reply app/models/piece.rb]: it is in a function and is being defined in the second line

Comment (NineDrop) [reply app/models/piece.rb:12]: line 4 puts this within the lat_check function,  line 6 defines start_x as the x_position of the piece

**Review by Claudwin — APPROVED** [review]
Went over the final commits and every case we tested gave us the expected outcome, if in the future we come across an edge case we didnt  anticipate we will work on it then but for now this works great!

**Pull request merged**